pkg/config/nodetreemodel/node.go

```diff
@@ -0,0 +1,367 @@
+}
+
+var _ LeafNode = (*leafNodeImpl)(nil)
+var _ Node = (*leafNodeImpl)(nil)
+
+// arrayNode represents a node with an ordered array of children
+
+type arrayNodeImpl struct {
+	nodes []Node
+}
+
+func newArrayNodeImpl(v []interface{}) (Node, error) {
+	nodes := make([]Node, 0, len(v))
+	for _, it := range v {
+		if n, ok := it.(Node); ok {
+			nodes = append(nodes, n)
+			continue
+		}
+		n, err := NewNode(it)
+		if err != nil {
+			return nil, err
+		}
+		nodes = append(nodes, n)
+	}
+	return &arrayNodeImpl{nodes: nodes}, nil
+}
+
+var _ ArrayNode = (*arrayNodeImpl)(nil)
+var _ Node = (*arrayNodeImpl)(nil)
+
+// node represents an arbitrary node of the tree
+
+type mapNodeImpl struct {
+	val map[string]interface{}
+	// remapCase maps each lower-case key to the original case. This
+	// enables GetChild to retrieve values using case-insensitive keys
+	remapCase map[string]string
+}
+
+func newMapNodeImpl(v map[string]interface{}) (Node, error) {
+	return &mapNodeImpl{val: v, remapCase: makeRemapCase(v)}, nil
+}
+
+var _ Node = (*mapNodeImpl)(nil)
+
+///////
+
+func isScalar(v interface{}) bool {
+	switch v.(type) {
+	case int, int8, int16, int32, int64:
+		return true
+	case uint, uint8, uint16, uint32, uint64:
+		return true
+	case bool, string, float32, float64, time.Time, time.Duration:
+		return true
+	default:
+		return false
+	}
+}
+
+// creates a map that converts keys from their lower-cased version to their original case
+func makeRemapCase(m map[string]interface{}) map[string]string {
+	remap := make(map[string]string)
+	for k := range m {
+		remap[strings.ToLower(k)] = k
+	}
+	return remap
+}
+
+func mapInterfaceToMapString(m map[interface{}]interface{}) map[string]interface{} {
+	res := make(map[string]interface{}, len(m))
+	for k, v := range m {
+		mk := ""
+		if str, ok := k.(string); ok {
+			mk = str
+		} else {
+			mk = fmt.Sprintf("%s", k)
+		}
+		res[mk] = v
+	}
+	return res
+}
+
+/////
+
+// GetChild returns the child node at the given case-insensitive key, or an error if not found
+func (n *mapNodeImpl) GetChild(key string) (Node, error) {
+	mkey := n.remapCase[strings.ToLower(key)]
+	child, found := n.val[mkey]
+	if !found {
+		return nil, ErrNotFound
+	}
+	// If the map is already storing a Node, return it
+	if n, ok := child.(Node); ok {
+		return n, nil
+	}
+	// Otherwise construct a new node
+	return NewNode(child)
+}
+
+// ChildrenKeys returns the list of keys of the children of the given node, if it is a map
+func (n *mapNodeImpl) ChildrenKeys() ([]string, error) {
+	mapkeys := make([]string, 0, len(n.val))
+	for key := range n.val {
+		mapkeys = append(mapkeys, key)
+	}
+	// map keys are iterated non-deterministically, sort them
+	slices.Sort(mapkeys)
+	return mapkeys, nil
+}
+
+// GetChild returns an error because array node does not have children accessible by name
+func (n *arrayNodeImpl) GetChild(string) (Node, error) {
+	return nil, fmt.Errorf("arrayNodeImpl.GetChild not implemented")
+}
```

Comment:
Here's the situation. We want `Node`'s in the tree to be able to have their `GetChild()` method return another `Node`. In order to do that, Node's in the tree have to implement the methods in this interface. An `ArrayNode` is one type that can show up in the tree, however it doesn't associate string-typed keys with its children. But we can't remove the `GetChild()` and `ChildrenKeys()` methods from `ArrayNode`, otherwise its no longer implementing the interface, which means it can't be returned by the `GetChild()` method of its parent.

For now, the only usages of these `Node`'s are (1) the `nodetreemodel` config, and (2) `UnmarshalKey`. Both of these callers will have to check-type `Node`'s to tell whether they're working with inner nodes, array nodes, or leaf nodes. So I think for now, it's best to leave these `ArrayNode` methods as unimplemented until we find a use case that requires a different behavior. Perhaps in the future we find that it's better to have `ArrayNode.GetChild()` return a `missingLeaf` instead (see missing.go), or we stringify the keys of `ChildrenKeys`, or we just modify the error message to be "not found" or something. I would rather wait for a use case that requires this before deciding now.